File: ai/rag/pipeline.py
"""Agentic RAG pipeline — docs + live web search + streaming."""
from __future__ import annotations

from datetime import datetime
from typing import AsyncIterator, List, Dict, Any, Optional

from langchain_core.messages import HumanMessage, AIMessage, SystemMessage

from . import vectorstore as vs
from .web_search import needs_web_search, search, format_results

# ── System prompts ────────────────────────────────────────────────────────────

_FORMAT_RULES = """
FORMATTING RULES (always follow):
- NEVER write a long paragraph. Always break into sections.
- Use **bold** for key terms, names, numbers.
- Use bullet points (- item) for lists of facts.
- Use numbered lists (1. item) for steps or rankings.
- Use ### heading for each section if answer has multiple topics.
- Use a markdown table if comparing multiple items.
- Short factual answers (date, name, number) → one line, no bullets needed.
- Code → always wrap in ```language blocks.
"""

_DOC_ONLY = """You are a precise knowledge assistant. The document context below is the user's own uploaded document (e.g. a resume, report, or notes).
Answer ONLY what the user asks using the document context.
{format_rules}
- When the user uses "my", "I", or "me", treat them as referring to the person described in the document.
- Extract the answer directly from the document text — do not say it is missing if the information is present.
- If the information is genuinely not in the document, say so in one line.

Document context:
{{context}}
""".format(format_rules=_FORMAT_RULES)

_WEB_ONLY = """You are a precise, up-to-date assistant with access to live web search results.
Answer ONLY what the user asks using the search results below.
{format_rules}
- ONLY use facts explicitly stated in the search results. Do NOT guess or infer.
- NEVER mix up dates — only report results matching today's date.
- For sports/matches: **Team A vs Team B**, venue, time — use bold and bullets.
- For weather/prices: state exact value with date.
- If results are unclear or conflicting, say so honestly.

Live web search results:
{{context}}
""".format(format_rules=_FORMAT_RULES)

_DOC_AND_WEB = """You are a precise knowledge assistant with document knowledge and live web data.
Answer ONLY what the user asks.
{format_rules}
- For personal, biographical, or document-specific questions (name, skills, experience, education, projects): use the Document context.
- For factual, current, or real-time questions (prices, exchange rates, news, weather, events): use the Web results and IGNORE the document.
- ONLY state facts explicitly in the sources — never guess.
- Cite web sources inline as (Source: URL).

Document context:
{{doc_context}}

Live web search results:
{{web_context}}
""".format(format_rules=_FORMAT_RULES)

_GENERAL = """You are a precise, helpful AI assistant. Answer ONLY what the user asks.
{format_rules}
- Do not pad, repeat, or add unrequested information.
- Match answer length to question complexity.
""".format(format_rules=_FORMAT_RULES)


# ── Context header ───────────────────────────────────────────────────────────

def _context_header() -> str:
    now = datetime.now()
    return (
        f"[ System context: "
        f"Current date: {now.strftime('%A, %d %B %Y')} | "
        f"Current time: {now.strftime('%I:%M %p')} ]\n\n"
    )


# ── Helpers ───────────────────────────────────────────────────────────────────

def _extract_text(content) -> str:
    """Safely extract string from LLM chunk (Gemini may return a list)."""
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        return "".join(
            b.get("text", "") if isinstance(b, dict) else str(b) for b in content
        )
    return str(content) if content else ""


def _format_docs(docs) -> str:
    parts = []
    for i, doc in enumerate(docs, 1):
        source = doc.metadata.get("source", "unknown")
        parts.append(f"[{i}] (source: {source})\n{doc.page_content}")
    return "\n\n---\n\n".join(parts)


def _build_messages(history: List[Dict], question: str, system_content: str):
    messages = [SystemMessage(content=system_content)]
    for turn in history[-10:]:
        messages.append(HumanMessage(content=turn["human"]))
        messages.append(AIMessage(content=turn["ai"]))
    messages.append(HumanMessage(content=question))
    return messages


# ── Main pipeline ─────────────────────────────────────────────────────────────

async def query_stream(
    question: str,
    provider,
    history: List[Dict],
    model: str | None = None,
    temperature: float = 0.3,
) -> AsyncIterator[str]:
    """
    Agentic routing:
      1. Check if question needs real-time data  → web search
      2. Check if knowledge base has relevant docs → RAG
      3. Combine both when available
      4. Fall back to general LLM if neither applies
    """
    doc_results = vs.similarity_search(question, k=6)
    has_docs    = bool(doc_results)

    # Always run web search for factual/current queries regardless of docs
    use_web     = needs_web_search(question)
    search_query = f"{question} {datetime.now().strftime('%d %B %Y')}" if use_web else question
    web_results  = search(search_query, max_results=3) if use_web else []

    has_web  = bool(web_results)

    sources: list[str] = []

    if has_docs and has_web:
        doc_context = _format_docs(doc_results)
        web_context = format_results(web_results)
        system_content = _DOC_AND_WEB.format(
            doc_context=doc_context, web_context=web_context
        )
        sources = (
            [d.metadata.get("source", "") for d in doc_results if d.metadata.get("source")]
            + [r["url"] for r in web_results if r.get("url")]
        )

    elif has_docs:
        system_content = _DOC_ONLY.format(context=_format_docs(doc_results))
        sources = [d.metadata.get("source", "") for d in doc_results if d.metadata.get("source")]

    elif has_web:
        system_content = _WEB_ONLY.format(context=format_results(web_results))
        sources = [r["url"] for r in web_results if r.get("url")]

    else:
        system_content = _GENERAL

    # Prepend current date/time to every system prompt
    system_content = _context_header() + system_content

    # Deduplicate sources
    sources = list(dict.fromkeys(s for s in sources if s))

    llm_kwargs: Dict[str, Any] = {"temperature": temperature}
    if model:
        llm_kwargs["model"] = model

    llm = provider.get_llm(**llm_kwargs)
    messages = _build_messages(history, question, system_content)

    async for chunk in llm.astream(messages):
        token = _extract_text(chunk.content)
        if token:
            yield token

    if sources:
        yield f"__SOURCES__:{','.join(sources)}"


def has_knowledge_base() -> bool:
    return vs.get_store() is not None

File: server/src/ai/rag/pipeline.js
// The RAG Pipeline ( The Brain )

import { HumanMessage, AIMessage, SystemMessage } from '@langchain/core/messages';
import * as vs from './vectorstore.js';
import { needsWebSearch, search, formatResults } from './webSearch.js';

// ── System prompts ─────────────────────────────────────────────────────────────

const FORMAT_RULES = `
FORMATTING RULES (always follow):
- NEVER write a long paragraph. Always break into sections.
- Use **bold** for key terms, names, numbers.
- Use bullet points (- item) for lists of facts.
- Use numbered lists (1. item) for steps or rankings.
- Use ### heading for each section if answer has multiple topics.
- Use a markdown table if comparing multiple items.
- Short factual answers (date, name, number) → one line, no bullets needed.
- Code → always wrap in \`\`\`language blocks.
`;

// Tells LLM to answer only from uploaded document context
const DOC_ONLY = (context) => `You are a precise knowledge assistant. The document context below is the user's own uploaded document.
Answer ONLY what the user asks using the document context.
${FORMAT_RULES}
- When the user uses "my", "I", or "me", treat them as referring to the person described in the document.
- Extract the answer directly from the document text.
- If the information is genuinely not in the document, say so in one line.

Document context:
${context}`;

// Tells LLM to answer only from live web search results
const WEB_ONLY = (context) => `You are a precise, up-to-date assistant with access to live web search results.
Answer ONLY what the user asks using the search results below.
${FORMAT_RULES}
- ONLY use facts explicitly stated in the search results. Do NOT guess or infer.
- NEVER mix up dates — only report results matching today's date.
- If results are unclear or conflicting, say so honestly.

Live web search results:
${context}`;

// Uses both; document for personal questions, web for current facts
const DOC_AND_WEB = (docCtx, webCtx) => `You are a precise knowledge assistant with document knowledge and live web data.
Answer ONLY what the user asks.
${FORMAT_RULES}
- For personal, biographical, or document-specific questions: use the Document context.
- For factual, current, or real-time questions: use the Web results and IGNORE the document.
- ONLY state facts explicitly in the sources — never guess.
- Cite web sources inline as (Source: URL).

Document context:
${docCtx}

Live web search results:
${webCtx}`;

// Plain AI assistant when no context found
const GENERAL = `You are a precise, helpful AI assistant. Answer ONLY what the user asks.
${FORMAT_RULES}
- Do not pad, repeat, or add unrequested information.
- Match answer length to question complexity.`;

// ── Helpers ────────────────────────────────────────────────────────────────────

// Prepends current date and time to every system prompt
function contextHeader() {
  const now  = new Date();
  const days = ['Sunday','Monday','Tuesday','Wednesday','Thursday','Friday','Saturday'];
  const mons = ['January','February','March','April','May','June','July','August','September','October','November','December'];
  const pad  = (n) => String(n).padStart(2, '0');
  const h    = now.getHours();
  const ampm = h >= 12 ? 'PM' : 'AM';
  const h12  = ((h % 12) || 12);
  return `[ System context: Current date: ${days[now.getDay()]}, ${pad(now.getDate())} ${mons[now.getMonth()]} ${now.getFullYear()} | Current time: ${pad(h12)}:${pad(now.getMinutes())} ${ampm} ]\n\n`;
}

function extractText(content) {
  if (typeof content === 'string') return content;
  if (Array.isArray(content)) return content.map(b => (typeof b === 'object' ? b.text || '' : String(b))).join('');
  return content ? String(content) : '';
}

function formatDocs(docs) {
  return docs.map((doc, i) => `[${i + 1}] (source: ${doc.metadata?.source || 'unknown'})\n${doc.pageContent}`).join('\n\n---\n\n');
}

// Builds the full LLM message list: system prompt + last 10 turns of history + current question
function buildMessages(history, question, systemContent) {
  const msgs = [new SystemMessage(systemContent)];
  for (const turn of history.slice(-10)) {
    msgs.push(new HumanMessage(turn.human));
    msgs.push(new AIMessage(turn.ai));
  }
  msgs.push(new HumanMessage(question));
  return msgs;
}

// ── Main pipeline ──────────────────────────────────────────────────────────────

export async function* queryStream(question, provider, history, model = null, temperature = 0.3, sessionId) {
  const docResults = await vs.similaritySearch(question, 6, 1.70, sessionId); //  Searches the vector store for document chunks related to the question (max 6 results)
  const hasDocs    = docResults.length > 0;

  const useWeb   = needsWebSearch(question);
  const now      = new Date();
  const dateStr  = `${now.getDate()} ${['January','February','March','April','May','June','July','August','September','October','November','December'][now.getMonth()]} ${now.getFullYear()}`;
  const webQuery = useWeb ? `${question} ${dateStr}` : question;
  const webResults = useWeb ? await search(webQuery, 3) : [];
  const hasWeb   = webResults.length > 0;

  let systemContent;
  let sources = [];

  // Picks the right system prompt based on what was found: DOC_ONLY, WEB_ONLY, DOC_AND_WEB, or GENERAL
  if (hasDocs && hasWeb) {
    systemContent = DOC_AND_WEB(formatDocs(docResults), formatResults(webResults));
    sources = [
      ...docResults.map(d => d.metadata?.source).filter(Boolean),
      ...webResults.map(r => r.url).filter(Boolean),
    ];
  } else if (hasDocs) {
    systemContent = DOC_ONLY(formatDocs(docResults));
    sources = docResults.map(d => d.metadata?.source).filter(Boolean);
  } else if (hasWeb) {
    systemContent = WEB_ONLY(formatResults(webResults));
    sources = webResults.map(r => r.url).filter(Boolean);
  } else {
    systemContent = GENERAL;
  }

  systemContent = contextHeader() + systemContent;
  sources = [...new Set(sources)];

  const llmOpts = { temperature };
  if (model) llmOpts.model = model;

  const llm      = provider.getLlm(llmOpts);
  const messages = buildMessages(history, question, systemContent);
  
  // Calls llm.stream(messages) — gets a token-by-token stream from OpenAI/Gemini/Llama, yields each token up to the controller
  const stream = await llm.stream(messages);
  for await (const chunk of stream) {
    const token = extractText(chunk.content);
    if (token) yield token;
  }

  if (sources.length) yield `__SOURCES__:${sources.join(',')}`; // After all tokens, yields __SOURCES__:url1,url2 as a single special token
}

export function hasKnowledgeBase(sessionId) {
  return vs.getStore(sessionId) !== null;
}
